src/borg/_hashindex.c

```diff
@@ -351,23 +359,168 @@ hashindex_read(PyObject *file_py, int permit_compact)
         goto fail_free_index;
     }
 
-    header = (HashHeader*) header_buffer.buf;
-    if(memcmp(header->magic, MAGIC, MAGIC_LEN)) {
+    header = (HashHeader1*) header_buffer.buf;
+    if(memcmp(header->magic, MAGIC1, MAGIC_LEN)) {
         PyErr_Format(PyExc_ValueError, "Unknown MAGIC in header");
         goto fail_release_header_buffer;
     }
 
     buckets_length = (Py_ssize_t)_le32toh(header->num_buckets) * (header->key_size + header->value_size);
-    if((Py_ssize_t)length != (Py_ssize_t)sizeof(HashHeader) + buckets_length) {
+    if((Py_ssize_t)length != (Py_ssize_t)sizeof(*header) + buckets_length) {
         PyErr_Format(PyExc_ValueError, "Incorrect file length (expected %zd, got %zd)",
-                     sizeof(HashHeader) + buckets_length, length);
+                     sizeof(*header) + buckets_length, length);
         goto fail_release_header_buffer;
     }
 
     index->num_entries = _le32toh(header->num_entries);
     index->num_buckets = _le32toh(header->num_buckets);
+    index->num_empty = -1;  // unknown, needs counting
     index->key_size = header->key_size;
     index->value_size = header->value_size;
+
+fail_release_header_buffer:
+    PyBuffer_Release(&header_buffer);
+fail_free_index:
+    if(PyErr_Occurred()) {
+        free(index);
+        index = NULL;
+    }
+fail_decref_header:
+    Py_DECREF(header_bytes);
+fail:
+    return index;
+}
+
+HashIndex *
+read_hashheader(PyObject *file_py)
+{
+    Py_ssize_t bytes_read, length, buckets_length;
+    Py_buffer header_buffer;
+    PyObject *header_bytes, *length_object, *tmp;
+    HashIndex *index = NULL;
+    HashHeader *header;
+
+    header_bytes = PyObject_CallMethod(file_py, "read", "n", (Py_ssize_t)sizeof(*header));
+    if(!header_bytes) {
+        assert(PyErr_Occurred());
+        goto fail;
+    }
+
+    bytes_read = PyBytes_Size(header_bytes);
+    if(PyErr_Occurred()) {
+        /* TypeError, not a bytes() object */
+        goto fail_decref_header;
+    }
+    if(bytes_read != sizeof(*header)) {
+        /* Truncated file */
+        /* Note: %zd is the format for Py_ssize_t, %zu is for size_t */
+        PyErr_Format(PyExc_ValueError, "Could not read header (expected %zu, but read %zd bytes)",
+                     sizeof(*header), bytes_read);
+        goto fail_decref_header;
+    }
+
+    /*
+     * Hash the header
+     * If the header is corrupted this bails before doing something stupid (like allocating 3.8 TB of memory)
+     */
+    tmp = PyObject_CallMethod(file_py, "hash_part", "s", "HashHeader");
+    Py_XDECREF(tmp);
+    if(PyErr_Occurred()) {
+        if(PyErr_ExceptionMatches(PyExc_AttributeError)) {
+            /* Be able to work with regular file objects which do not have a hash_part method. */
+            PyErr_Clear();
+        } else {
+            goto fail_decref_header;
+        }
+    }
```

Comment:
I don't know how this works (I took a brief look at `file_integrity.py`, but I'm not awake enough yet to follow it completely), so the obvious question: The hash is not in a place where it creates an offset for the post-header data, right? I.e. this file integrity thing does not break the alignment, does it?

Reply:
no, the integrity data is in a separate file.